### PR TITLE
Model: Unify PartialEq

### PR DESCRIPTION
**Pull request by hunger**
Implement PartialEq for Rc<Model> and use that for ModelHandle as well
as Value<Model>

Wrap a Rc<Model> into a SharedModel and implement PartialEq for that.

Wrap Option<SharedModel> into the ModuleHandle as well as Value::Model.

### CODE REVIEW

File: sixtyfps_runtime/corelib/model.rs [hunk 1]
```diff
@@ -205,14 +205,49 @@ impl Model for bool {
     }
 }
 
+/// Properties of type array in the .60 language are represented as
+/// an [`Option`] of an [`Rc`] of something implemented the [`Model`] trait
+#[derive(derive_more::Deref, derive_more::DerefMut)]
+pub struct SharedModel<T>(pub Rc<dyn Model<Data = T>>);
```
Comment (tronical) [inline sixtyfps_runtime/corelib/model.rs:211]: Should this be repr(transparent)?

Comment (hunger) [reply sixtyfps_runtime/corelib/model.rs:211]: Makes probably sense, yes.

File: sixtyfps_runtime/corelib/model.rs [hunk 2]
```diff
@@ -231,23 +267,16 @@ impl<T> core::cmp::PartialEq for ModelHandle<T> {
...
-            (None, Some(_)) => false,
-            (Some(_), None) => false,
-            (Some(a), Some(b)) => Rc::ptr_eq(a, b),
+            (Some(a), Some(b)) => a == b,
+            _ => false,
         }
     }
 }
 
-impl<T> From<Rc<dyn Model<Data = T>>> for ModelHandle<T> {
```
Comment (tronical) [inline sixtyfps_runtime/corelib/model.rs:241]: Just curious, why was this removed?

Comment (hunger) [reply sixtyfps_runtime/corelib/model.rs:241]: I did not need it to build, but then I apparently did not build everything somehow.

At least it did build and test just fine on my local machine and fails to build here.

**Comment by tronical** [comment]
I think that while this change is source incompatible, it is good - especially the centralization of the ptr_eq.

This sheds a bit of light on the fact that there are quite some ways of constructing a modelhandle :). Is that why you removed the From?

**Comment by hunger** [comment]
@tronical: Yes, this definitely needs some thinking before it can go in.

Is there some documentation on the cpp_cmake parts somewhere? Looks like I broke those.

**Comment by tronical** [comment]
It seems to fail deep in the interpreter api through the cmake build. I’ll try to write something about how that plays together on Monday. Sorry about this can of worms :)

**Comment by tronical** [comment]
I think the heart of the error is this:

```
Error: api/sixtyfps-cpp/generated_include/sixtyfps_interpreter_internal.h:177:9: error: ‘SharedModel’ does not name a type
```

which indicates that the new `Value` variant holding `SharedModel` is failing because `SharedModel` is not a type known in C++.

**Comment by tronical** [comment]
> I think the heart of the error is this:
> 
> ```
> Error: api/sixtyfps-cpp/generated_include/sixtyfps_interpreter_internal.h:177:9: error: ‘SharedModel’ does not name a type
> ```
> 
> which indicates that the new `Value` variant holding `SharedModel` is failing because `SharedModel` is not a type known in C++.

Ah, but `Value` should not be exposed in the generated headers in the first place.

**Comment by hunger** [comment]
@tronical: Ok, just did a fresh push, let's see what happens.

File: sixtyfps_runtime/corelib/model.rs [hunk 3]
```diff
@@ -205,14 +205,51 @@ impl Model for bool {
...
+#[derive(derive_more::Deref, derive_more::DerefMut)]
+pub struct SharedModel<T>(pub Rc<dyn Model<Data = T>>);
+
+impl<T> std::fmt::Debug for SharedModel<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "SharedModel(dyn Model)")
+    }
+}
+
+impl<T> Clone for SharedModel<T> {
```
Comment (tronical) [inline sixtyfps_runtime/corelib/model.rs:219]: I suppose this can also be achieved using `#[derive(Clone)]`?

Comment (tronical) [reply sixtyfps_runtime/corelib/model.rs:219]: Oh, I see, same in ModelHandle, so not relevant for this change set :)

File: sixtyfps_runtime/corelib/model.rs (hunk 2 above)
Comment (tronical) [reply sixtyfps_runtime/corelib/model.rs:241]: IMO is worth keeping this public API unless there’s a good reason to remove it (such as clippy complaining). IIRC initially the Handle had just the Rc but we had the add the Option, and then this impl was added to make the API easier to use.

**Comment by tronical** [comment]
> @tronical: Ok, just did a fresh push, let's see what happens.

Yay, it’s green now. Only two things left IMO:

(1) typo in the commit message (Hanlde-> Handle)
(2) either break public API but mention in the commit message why (or edit changelog), or keep it - regarding the From impl

File: sixtyfps_runtime/corelib/model.rs (hunk 3 above)
Comment (ogoffart) [reply sixtyfps_runtime/corelib/model.rs:219]: the problem with #[derive(Clone)] is that it adds a `where T: Clone` which we don't want.

File: sixtyfps_runtime/corelib/model.rs [hunk 4]
```diff
@@ -205,14 +205,51 @@ impl Model for bool {
...
+}
+
 /// Properties of type array in the .60 language are represented as
 /// an [`Option`] of an [`Rc`] of something implemented the [`Model`] trait
-#[derive(derive_more::Deref, derive_more::DerefMut, derive_more::From, derive_more::Into)]
-pub struct ModelHandle<T>(pub Option<Rc<dyn Model<Data = T>>>);
+#[derive(
+    PartialEq, derive_more::Deref, derive_more::DerefMut, derive_more::From, derive_more::Into,
+)]
+pub struct ModelHandle<T>(pub Option<SharedModel<T>>);
```
Comment (ogoffart) [inline sixtyfps_runtime/corelib/model.rs:248]: i wonder if there is a way to remove the inner `pub`

Anyway, this is a public API breakage.

File: sixtyfps_runtime/corelib/model.rs [hunk 5]
```diff
@@ -205,14 +205,51 @@ impl Model for bool {
...
+
+impl<T> core::cmp::PartialEq for SharedModel<T> {
+    #[allow(clippy::vtable_address_comparisons)]
+    // We think vtable_address_comparisons are safe here!
+    // Rc::ptr_eq will compare a fat pointer consisting of a `vtable` and a `self` pointer.
+    //
+    // We can get:
+    // * false positives: Two different `Model`s get considered the same, even though they are not.
+    //   For this the `self` pointer need to point to the same place. That can only happen if
+    //   `Self` is `()`, which does not make sense for a `Model`. So we ignore this case.
```
Comment (ogoffart) [inline sixtyfps_runtime/corelib/model.rs:233]: Self could very well be `()`.

We could imagine for example a  `impl Model for ()` that always returns 0 elements.

or a  `struct MyStaticModel;  impl Model for MyStaticModel { ... }`.

But anyway, this cannot happen that two different model are considered the same, because the Rc adds an allocation for the refcounts and always allocates. so the data pointer will always be different anyway.

File: sixtyfps_runtime/corelib/model.rs [hunk 6]
```diff
@@ -205,14 +205,51 @@ impl Model for bool {
...
+    // * false negatives: The same `Model` gets considered to be two different `Model`s.
+    //   In this case the `vtable` pointer must differ, e.g. because two crates created separate
+    //   instances of the vtable for the `Model` trait.
+    //   This will trigger some unnecessary rendering, but is benign otherwise.
```
Comment (ogoffart) [inline sixtyfps_runtime/corelib/model.rs:237]: yes, this is correct.


(but now that i think about it, because this is Rc, a data pointer comparison would also work as intended.)

File: sixtyfps_runtime/interpreter/api.rs [hunk 7]
```diff
@@ -152,7 +152,7 @@ impl PartialEq for Value {
             Value::Bool(lhs) => matches!(other, Value::Bool(rhs) if lhs == rhs),
             Value::Image(lhs) => matches!(other, Value::Image(rhs) if lhs == rhs),
             Value::Array(lhs) => matches!(other, Value::Array(rhs) if lhs == rhs),
-            Value::Model(lhs) => matches!(other, Value::Model(rhs) if Rc::ptr_eq(lhs, rhs)),
+            Value::Model(lhs) => matches!(other, Value::Model(rhs) if lhs == rhs),
```
Comment (ogoffart) [inline sixtyfps_runtime/interpreter/api.rs:155]: I actually wonder if this was not the only reason we implemented PartialEq manually.
We could just use #[derive(PartialEq)] instead, now.

**Review by ogoffart — COMMENTED** [review]
I've tagged the change as `breaking_change` since this change the public API.

The changes here are overall good. But we haven't decide if the next release of the `sixtyfps` crate will be 0.1.1 or 0.2.

I guess we should wait a bit to see if we have other reasons to break the source compatibility before merging this.

Also changes to public API deserve a changelog entry.

(It would be nice if there was a CI check that could tell if we are breaking semver or touching the public API)

File: sixtyfps_runtime/corelib/model.rs (hunk 5 above)
Comment (hunger) [reply sixtyfps_runtime/corelib/model.rs:233]: @ogoffart: You are way deeper in this rabbit hole than me, so did I understand this correctly?
 
 But I think you are right, this lint might actually be wrong: I went ahead and filed https://github.com/rust-lang/rust-clippy/issues/7461 with my current understanding. I am pretty sure I missed something, but maybe some improvement to the explanation of that clippy lint will fall out of this:-)

File: sixtyfps_runtime/interpreter/api.rs (hunk 7 above)
Comment (hunger) [reply sixtyfps_runtime/interpreter/api.rs:155]: I can do that, but would prefer to do that in a separate patch.

File: sixtyfps_runtime/corelib/model.rs (hunk 2 above)
Comment (hunger) [reply sixtyfps_runtime/corelib/model.rs:241]: I will add this back in in the next push.

File: sixtyfps_runtime/corelib/model.rs (hunk 4 above)
Comment (hunger) [reply sixtyfps_runtime/corelib/model.rs:248]: True, I added something about this to the CHANGELOG now.

File: sixtyfps_runtime/corelib/model.rs (hunk 5 above)
Comment (hunger) [reply sixtyfps_runtime/corelib/model.rs:233]: I updated the text, I hope it is a bit better now.
